spec.bs

```diff
@@ -403,6 +423,139 @@ Each {{AppHistory}} object has an associated <dfn for="AppHistory">navigate meth
+
+  1. Let |key| be [=this=]'s [=AppHistory/entry list=][[=this=]'s [=AppHistory/current index=] + 1]'s [=AppHistoryEntry/session history entry=]'s [=session history entry/app history key=].
+
+  1. Return the result of [=performing an app history traversal=] given [=this=], |key|, and |options|.
+</div>
+
+<div algorithm>
+  <p class="advisement">The following algorithm is specified in terms of the <a href="https://github.com/whatwg/html/pull/6315">session history rewrite pull request</a> against the HTML Standard, because the existing session history traversal infrastructure is broken enough that it's hard to build on. It is expected to track that work as it continues.</p>
+
+  To <dfn>perform an app history traversal</dfn> given an {{AppHistory}} object |appHistory|, a string |key|, and an {{AppHistoryNavigationOptions}} |options|:
+
+  1. If |appHistory|'s [=relevant global object=]'s [=associated Document=] is not [=Document/fully active=], then return [=a promise rejected with=] an "{{InvalidStateError}}" {{DOMException}}.
+
+  1. If |appHistory|'s [=relevant global object=]'s [=associated Document=]'s <a spec="HTML">unload counter</a> is greater than 0, then return [=a promise rejected with=] an "{{InvalidStateError}}" {{DOMException}}.
+
+  1. If |appHistory|'s [=AppHistory/entry list=][|appHistory|'s [=AppHistory/current index=]]'s [=AppHistoryEntry/session history entry=]'s [=session history entry/app history key=] equals |key|, then return [=a promise resolved with=] undefined.
+
+  1. Let |navigateInfo| be |options|["{{AppHistoryNavigationOptions/navigateInfo}}"] if it [=map/exists=], or undefined otherwise.
+
+  1. Let |navigable| be |appHistory|'s [=relevant global object=]'s [=Window/browsing context=]'s [=browsing context/containing navigable=].
+
+  1. Let |traversable| be |navigable|'s [=navigable/traversable navigable=].
+
+  1. Let |initiatorBC| be |appHistory|'s [=relevant global object=]'s [=Window/browsing context=].
+
+  1. Let |promise| be [=a new promise=] created in |appHistory|'s [=relevant Realm=].
+
+  1. [=parallel queue/Enqueue the following steps=] on |traversable|'s [=traversable navigable/session history traversal queue=]:
+
+    1. Let |navigableEntries| be the result of [=navigable/getting the session history entries=] given |navigable|.
+
+    1. Let |targetEntry| be the [=session history entry=] in |navigableEntries| whose [=session history entry/app history key=] equals |key|. If no such entry exists, then:
+
+      1. [=Reject=] |promise| with an "{{InvalidStateError}}" {{DOMException}}.
+
+      1. Abort these steps.
+
+      <p class="note">This can occur if the |appHistory| object's view of session history is outdated, which can happen for brief periods while all the relevant threads and processes are being synchronized in reaction to a history change (such as the user clearing their history).
+
+    1. Assert: |targetEntry| is not |navigable|'s [=navigable/active session history entry=].
```

Comment:
Note that we synchronously returned a promise fulfilled with undefined if you do `appHistory.goTo(appHistory.current.key)`. Nevertheless, I'm not sure if this can be an assert or whether it needs to be checked again in this parallel queue.

My intuition is that an insert is safe because it seems impossible for a frame to get such an outdated view of its own active entry. And, we plan to abort any traversals if any other traversals start (not yet specced), so you couldn't do something like `const k = appHistory.entries()[appHistory.current.index - 1].key; appHistory.goTo(k); appHistory.goTo(k)` which would queue up two such traversals where the second one would hit this. But, maybe I'm not being devious enough.